File: loadVoteDataToDatabase.js
require('dotenv').config();
const sql = require('mssql');
const fs = require('fs').promises;
const path = require('path');

// Database configuration
const config = {
  user: process.env.DB_USERNAME,
  password: process.env.DB_PASSWORD,
  server: process.env.DB_SERVER,
  database: process.env.DB_DATABASE,
  options: {
    encrypt: true,
    trustServerCertificate: false
  }
};

// Table schema for plenary session vote data
const voteTableSchema = `
  CREATE TABLE assembly_plenary_session_vote (
    id INT IDENTITY(1,1) PRIMARY KEY,
    HG_NM NVARCHAR(100),
    HJ_NM NVARCHAR(100),
    POLY_NM NVARCHAR(100),
    ORIG_NM NVARCHAR(100),
    MEMBER_NO NVARCHAR(50),
    POLY_CD NVARCHAR(50),
    ORIG_CD NVARCHAR(50),
    VOTE_DATE NVARCHAR(50),
    BILL_NO NVARCHAR(50),
    BILL_NAME NVARCHAR(500),
    BILL_ID NVARCHAR(100),
    LAW_TITLE NVARCHAR(500),
    CURR_COMMITTEE NVARCHAR(200),
    RESULT_VOTE_MOD NVARCHAR(50),
    DEPT_CD NVARCHAR(50),
    CURR_COMMITTEE_ID NVARCHAR(50),
    DISP_ORDER INT,
    BILL_URL NVARCHAR(1000),
    BILL_NAME_URL NVARCHAR(1000),
    SESSION_CD INT,
    CURRENTS_CD INT,
    AGE INT,
    MONA_CD NVARCHAR(50),
    created_at DATETIME2 DEFAULT GETDATE()
  )
`;

async function createVoteTable(pool) {
  console.log('Creating assembly_plenary_session_vote table...');
  try {
    // Check if table exists
    const checkResult = await pool.request()
      .input('tableName', sql.NVarChar, 'assembly_plenary_session_vote')
      .query(`
        SELECT COUNT(*) as count 
        FROM INFORMATION_SCHEMA.TABLES 
        WHERE TABLE_NAME = @tableName
      `);
    
    if (checkResult.recordset[0].count === 0) {
      await pool.request().query(voteTableSchema);
      console.log('Table assembly_plenary_session_vote created.');
    } else {
      console.log('Table assembly_plenary_session_vote already exists.');
    }
  } catch (error) {
    console.error('Error creating vote table:', error);
    throw error;
  }
}

async function insertVoteData(pool, voteItem, originalBillId, originalAge, apiStatus) {
  try {
    await pool.request()
      .input('HG_NM', sql.NVarChar(100), voteItem.HG_NM)
      .input('HJ_NM', sql.NVarChar(100), voteItem.HJ_NM)
      .input('POLY_NM', sql.NVarChar(100), voteItem.POLY_NM)
      .input('ORIG_NM', sql.NVarChar(100), voteItem.ORIG_NM)
      .input('MEMBER_NO', sql.NVarChar(50), voteItem.MEMBER_NO)
      .input('POLY_CD', sql.NVarChar(50), voteItem.POLY_CD)
      .input('ORIG_CD', sql.NVarChar(50), voteItem.ORIG_CD)
      .input('VOTE_DATE', sql.NVarChar(50), voteItem.VOTE_DATE)
      .input('BILL_NO', sql.NVarChar(50), voteItem.BILL_NO)
      .input('BILL_NAME', sql.NVarChar(500), voteItem.BILL_NAME)
      .input('BILL_ID', sql.NVarChar(100), voteItem.BILL_ID)
      .input('LAW_TITLE', sql.NVarChar(500), voteItem.LAW_TITLE)
      .input('CURR_COMMITTEE', sql.NVarChar(200), voteItem.CURR_COMMITTEE)
      .input('RESULT_VOTE_MOD', sql.NVarChar(50), voteItem.RESULT_VOTE_MOD)
      .input('DEPT_CD', sql.NVarChar(50), voteItem.DEPT_CD)
      .input('CURR_COMMITTEE_ID', sql.NVarChar(50), voteItem.CURR_COMMITTEE_ID)
      .input('DISP_ORDER', sql.Int, voteItem.DISP_ORDER)
      .input('BILL_URL', sql.NVarChar(1000), voteItem.BILL_URL)
      .input('BILL_NAME_URL', sql.NVarChar(1000), voteItem.BILL_NAME_URL)
      .input('SESSION_CD', sql.Int, voteItem.SESSION_CD)
      .input('CURRENTS_CD', sql.Int, voteItem.CURRENTS_CD)
      .input('AGE', sql.Int, voteItem.AGE)
      .input('MONA_CD', sql.NVarChar(50), voteItem.MONA_CD)
      .query(`
        INSERT INTO assembly_plenary_session_vote (
          HG_NM, HJ_NM, POLY_NM, ORIG_NM, MEMBER_NO, POLY_CD, ORIG_CD, VOTE_DATE,
          BILL_NO, BILL_NAME, BILL_ID, LAW_TITLE, CURR_COMMITTEE, RESULT_VOTE_MOD,
          DEPT_CD, CURR_COMMITTEE_ID, DISP_ORDER, BILL_URL, BILL_NAME_URL,
          SESSION_CD, CURRENTS_CD, AGE, MONA_CD
        ) VALUES (
          @HG_NM, @HJ_NM, @POLY_NM, @ORIG_NM, @MEMBER_NO, @POLY_CD, @ORIG_CD, @VOTE_DATE,
          @BILL_NO, @BILL_NAME, @BILL_ID, @LAW_TITLE, @CURR_COMMITTEE, @RESULT_VOTE_MOD,
          @DEPT_CD, @CURR_COMMITTEE_ID, @DISP_ORDER, @BILL_URL, @BILL_NAME_URL,
          @SESSION_CD, @CURRENTS_CD, @AGE, @MONA_CD
        )
      `);
  } catch (error) {
    console.error('Error inserting vote record:', error.message);
    throw error;
  }
}

async function loadAPIResultsData() {
  try {
    console.log('Loading API results data...');
    
    let apiData;
    let dataSource;
    
    // Try to load main results file first
    try {
      const mainPath = path.join(__dirname, 'assembly_bills_api_results.json');
      const content = await fs.readFile(mainPath, 'utf8');
      apiData = JSON.parse(content);
      dataSource = 'assembly_bills_api_results.json';
      console.log('Loaded data from main results file');
    } catch (error) {
      // Fallback to temp file
      try {
        const tempPath = path.join(__dirname, 'assembly_bills_api_results_temp.json');
        const content = await fs.readFile(tempPath, 'utf8');
        apiData = JSON.parse(content);
        dataSource = 'assembly_bills_api_results_temp.json';
        console.log('Loaded data from temp results file');
      } catch (tempError) {
        throw new Error('Neither main nor temp API results file found');
      }
    }
    
    if (!apiData.results || !Array.isArray(apiData.results)) {
      throw new Error('No results array found in API data');
    }
    
    console.log(`Data source: ${dataSource}`);
    console.log(`Found ${apiData.results.length} API result records`);
    
    return apiData;
    
  } catch (error) {
    console.error('Error loading API results data:', error);
    throw error;
  }
}

async function processAPIResults(pool, apiData) {
  console.log('Processing API results for vote data...');
  
  let processedCount = 0;
  let successfulInserts = 0;
  let voteRecordsFound = 0;

  // let flag = false;
  
  for (const result of apiData.results) {
    processedCount++;

    // TODO 오류 발생한 BILL_ID 기준으로 작업 재시작
    // console.log('result.BILL_ID', result.BILL_ID)
    // if (result.BILL_ID == 'PRC_W1R7Y0O8M0J9E1W7E4M3U4Q1K2V5X5') flag = true
    // if (flag == false) continue;
    
    if (processedCount % 100 === 0) {
      console.log(`Processed ${processedCount}/${apiData.results.length} records...`);
    }
    
    // Only process successful API responses
    if (result.status !== 'success' || !result.api_response) {
      continue;
    }
    
    try {
      // Check if api_response has the expected structure
      if (!Array.isArray(result.api_response) || result.api_response.length < 2) {
        continue;
      }
      
      // Get the second element (index 1) from api_response array
      const voteData = result.api_response[1];
      
      if (!voteData || !Array.isArray(voteData.row)) {
        continue;
      }
      
      // Process each vote record in the row array
      for (const voteItem of voteData.row) {
        voteRecordsFound++;
        
        try {
          await insertVoteData(
            pool, 
            voteItem, 
            result.BILL_ID, 
            result.AGE, 
            result.status
          );
          successfulInserts++;
        } catch (insertError) {
          console.error(`Error inserting vote record for BILL_ID ${result.BILL_ID}:`, insertError.message);
        }
      }
      
    } catch (error) {
      console.error(`Error processing result for BILL_ID ${result.BILL_ID}:`, error.message);
    }
  }
  
  console.log(`\nProcessing completed:`);
  console.log(`- Total API results processed: ${processedCount}`);
  console.log(`- Vote records found: ${voteRecordsFound}`);
  console.log(`- Successful database insertions: ${successfulInserts}`);
  
  return {
    processedCount,
    voteRecordsFound,
    successfulInserts
  };
}

async function main() {
  let pool;
  
  try {
    console.log('Connecting to Azure SQL Database...');
    pool = await sql.connect(config);
    console.log('Connected successfully!');
    
    // Create vote table
    await createVoteTable(pool);
    
    // Load API results data
    const apiData = await loadAPIResultsData();
    
    // Process and insert vote data
    const results = await processAPIResults(pool, apiData);
    
    console.log('\nData loading completed successfully!');
    console.log(`Summary: ${results.successfulInserts} vote records inserted from ${results.voteRecordsFound} found records`);
    
  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  } finally {
    if (pool) {
      await pool.close();
      console.log('Database connection closed.');
    }
  }
}

if (require.main === module) {
  main();
}

module.exports = { main, loadAPIResultsData, processAPIResults };

File: loadVoteDataToDatabase.py
import os
import json
import pyodbc
from datetime import datetime
from pathlib import Path
from typing import List, Dict, Any, Optional
from dotenv import load_dotenv

class VoteDataLoader:
    def __init__(self):
        # Load environment variables
        load_dotenv()
        
        self.base_dir = Path(__file__).parent
        
        # Database configuration
        self.db_config = {
            'server': os.getenv('DB_SERVER'),
            'database': os.getenv('DB_DATABASE'),
            'username': os.getenv('DB_USERNAME'),
            'password': os.getenv('DB_PASSWORD'),
            'driver': '{ODBC Driver 17 for SQL Server}'  # or '{SQL Server}'
        }
        
        # Table schema for plenary session vote data
        self.vote_table_schema = """
            CREATE TABLE assembly_plenary_session_vote (
                id INT IDENTITY(1,1) PRIMARY KEY,
                HG_NM NVARCHAR(100),
                HJ_NM NVARCHAR(100),
                POLY_NM NVARCHAR(100),
                ORIG_NM NVARCHAR(100),
                MEMBER_NO NVARCHAR(50),
                POLY_CD NVARCHAR(50),
                ORIG_CD NVARCHAR(50),
                VOTE_DATE NVARCHAR(50),
                BILL_NO NVARCHAR(50),
                BILL_NAME NVARCHAR(500),
                BILL_ID NVARCHAR(100),
                LAW_TITLE NVARCHAR(500),
                CURR_COMMITTEE NVARCHAR(200),
                RESULT_VOTE_MOD NVARCHAR(50),
                DEPT_CD NVARCHAR(50),
                CURR_COMMITTEE_ID NVARCHAR(50),
                DISP_ORDER INT,
                BILL_URL NVARCHAR(1000),
                BILL_NAME_URL NVARCHAR(1000),
                SESSION_CD INT,
                CURRENTS_CD INT,
                AGE INT,
                MONA_CD NVARCHAR(50),
                created_at DATETIME2 DEFAULT GETDATE()
            )
        """

    def get_connection(self):
        """Create database connection"""
        connection_string = (
            f"DRIVER={self.db_config['driver']};"
            f"SERVER={self.db_config['server']};"
            f"DATABASE={self.db_config['database']};"
            f"UID={self.db_config['username']};"
            f"PWD={self.db_config['password']};"
            "Encrypt=yes;TrustServerCertificate=no;"
        )
        return pyodbc.connect(connection_string)

    def create_vote_table(self, cursor) -> None:
        """Create assembly_plenary_session_vote table if it doesn't exist"""
        print('Creating assembly_plenary_session_vote table...')
        
        try:
            # Check if table exists
            cursor.execute("""
                SELECT COUNT(*) 
                FROM INFORMATION_SCHEMA.TABLES 
                WHERE TABLE_NAME = 'assembly_plenary_session_vote'
            """)
            
            if cursor.fetchone()[0] == 0:
                cursor.execute(self.vote_table_schema)
                print('Table assembly_plenary_session_vote created.')
            else:
                print('Table assembly_plenary_session_vote already exists.')
        except Exception as error:
            print(f'Error creating vote table: {error}')
            raise error

    def insert_vote_data(self, cursor, vote_item: Dict[str, Any], original_bill_id: str, 
                        original_age: str, api_status: str) -> None:
        """Insert a single vote record into the database"""
        try:
            cursor.execute("""
                INSERT INTO assembly_plenary_session_vote (
                    HG_NM, HJ_NM, POLY_NM, ORIG_NM, MEMBER_NO, POLY_CD, ORIG_CD, VOTE_DATE,
                    BILL_NO, BILL_NAME, BILL_ID, LAW_TITLE, CURR_COMMITTEE, RESULT_VOTE_MOD,
                    DEPT_CD, CURR_COMMITTEE_ID, DISP_ORDER, BILL_URL, BILL_NAME_URL,
                    SESSION_CD, CURRENTS_CD, AGE, MONA_CD
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """, (
                vote_item.get('HG_NM'),
                vote_item.get('HJ_NM'),
                vote_item.get('POLY_NM'),
                vote_item.get('ORIG_NM'),
                vote_item.get('MEMBER_NO'),
                vote_item.get('POLY_CD'),
                vote_item.get('ORIG_CD'),
                vote_item.get('VOTE_DATE'),
                vote_item.get('BILL_NO'),
                vote_item.get('BILL_NAME'),
                vote_item.get('BILL_ID'),
                vote_item.get('LAW_TITLE'),
                vote_item.get('CURR_COMMITTEE'),
                vote_item.get('RESULT_VOTE_MOD'),
                vote_item.get('DEPT_CD'),
                vote_item.get('CURR_COMMITTEE_ID'),
                vote_item.get('DISP_ORDER'),
                vote_item.get('BILL_URL'),
                vote_item.get('BILL_NAME_URL'),
                vote_item.get('SESSION_CD'),
                vote_item.get('CURRENTS_CD'),
                vote_item.get('AGE'),
                vote_item.get('MONA_CD')
            ))
        except Exception as error:
            print(f'Error inserting vote record: {error}')
            raise error

    def load_api_results_data(self) -> Dict[str, Any]:
        """Load API results data from JSON file"""
        try:
            print('Loading API results data...')
            
            # Try to load main results file first
            main_path = self.base_dir / 'assembly_bills_api_results.json'
            if main_path.exists():
                with open(main_path, 'r', encoding='utf-8') as f:
                    api_data = json.load(f)
                data_source = 'assembly_bills_api_results.json'
                print('Loaded data from main results file')
            else:
                # Fallback to temp file
                temp_path = self.base_dir / 'assembly_bills_api_results_temp.json'
                if temp_path.exists():
                    with open(temp_path, 'r', encoding='utf-8') as f:
                        api_data = json.load(f)
                    data_source = 'assembly_bills_api_results_temp.json'
                    print('Loaded data from temp results file')
                else:
                    raise Exception('Neither main nor temp API results file found')
            
            if 'results' not in api_data or not isinstance(api_data['results'], list):
                raise Exception('No results array found in API data')
            
            print(f'Data source: {data_source}')
            print(f'Found {len(api_data["results"])} API result records')
            
            return api_data
            
        except Exception as error:
            print(f'Error loading API results data: {error}')
            raise error

    def process_api_results(self, cursor, api_data: Dict[str, Any]) -> Dict[str, int]:
        """Process API results for vote data"""
        print('Processing API results for vote data...')
        
        processed_count = 0
        successful_inserts = 0
        vote_records_found = 0
        
        for result in api_data['results']:
            processed_count += 1
            
            if processed_count % 100 == 0:
                print(f'Processed {processed_count}/{len(api_data["results"])} records...')
            
            # Only process successful API responses
            if result.get('status') != 'success' or not result.get('api_response'):
                continue
            
            try:
                # Check if api_response has the expected structure
                if not isinstance(result['api_response'], list) or len(result['api_response']) < 2:
                    continue
                
                # Get the second element (index 1) from api_response array
                vote_data = result['api_response'][1]
                
                if not vote_data or not isinstance(vote_data.get('row'), list):
                    continue
                
                # Process each vote record in the row array
                for vote_item in vote_data['row']:
                    vote_records_found += 1
                    
                    try:
                        self.insert_vote_data(
                            cursor,
                            vote_item,
                            result['BILL_ID'],
                            result['AGE'],
                            result['status']
                        )
                        successful_inserts += 1
                    except Exception as insert_error:
                        print(f'Error inserting vote record for BILL_ID {result["BILL_ID"]}: {insert_error}')
                
            except Exception as error:
                print(f'Error processing result for BILL_ID {result["BILL_ID"]}: {error}')
        
        print(f'\nProcessing completed:')
        print(f'- Total API results processed: {processed_count}')
        print(f'- Vote records found: {vote_records_found}')
        print(f'- Successful database insertions: {successful_inserts}')
        
        return {
            'processed_count': processed_count,
            'vote_records_found': vote_records_found,
            'successful_inserts': successful_inserts
        }

    def run(self) -> None:
        """Main execution method"""
        try:
            print('Connecting to SQL Server Database...')
            connection = self.get_connection()
            cursor = connection.cursor()
            print('Connected successfully!')
            
            # Create vote table
            self.create_vote_table(cursor)
            
            # Load API results data
            api_data = self.load_api_results_data()
            
            # Process and insert vote data
            results = self.process_api_results(cursor, api_data)
            
            # Commit all changes
            connection.commit()
            
            print('\nData loading completed successfully!')
            print(f'Summary: {results["successful_inserts"]} vote records inserted from {results["vote_records_found"]} found records')
            
        except Exception as error:
            print(f'Error: {error}')
            if 'connection' in locals():
                connection.rollback()
            raise
        finally:
            if 'connection' in locals():
                connection.close()
                print('Database connection closed.')

def main():
    """Main function to run the vote data loader"""
    loader = VoteDataLoader()
    loader.run()

if __name__ == "__main__":
    main()